### PR TITLE
Non-blocking socket support/better handling of HTTP/1.1 connections

### PR DESCRIPTION
**Pull request by hexaclock**
* **What kind of change does this PR introduce?**
  - [X] bug fix
  - [ ] feature
  - [ ] docs update
  - [ ] tests/coverage improvement
  - [ ] refactoring
  - [ ] other

* **What is the related issue number (starting with `#`)**
#91

* **What is the current behavior?** (You can also link to an open issue here)
#91 

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

* **Checklist**:

  - [x] I think the code is well written
  - [ ] I wrote [good commit messages][1]
  - [ ] I have [squashed related commits together][2] after the changes have been approved
  - [ ] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [X] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/117)

### CODE REVIEW

**Comment by webknjaz** [comment]
@hexaclock why didn't you use a feature-branch in your fork for this PR?

**Comment by hexaclock** [comment]
Sorry about that. Dealing with Bash on Windows :)

I'll clean up the rest throughout the week/weekend.

**Comment by webknjaz** [comment]
This pull request **introduces 3 alerts** when merging c3bd269af5feb59c5b431ecfbb2383be7b3f1fcb into f5093ce18f3ab7c98d3dffa6556a2356f6079ff4 - [view on LGTM.com](https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-4f3dfe2b9989b7a41e3eacde31b659323741e106)

**new alerts:**

* 1 for First argument of a method is not named &#39;self&#39;
* 1 for Non\-iterable used in for loop
* 1 for Wrong number of arguments in a call

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
Oh, and `test_HTTP11_Timeout` is failing, you'll have to fix it.

**Comment by hexaclock** [comment]
Yep will do. Regarding `test_HTTP11_Timeout`, I'll probably need to add a new config option to explicitly enable non-blocking sockets so that we can cleanly differentiate between the different behaviors.

I've been unexpectedly busy these past few weeks, but I'm hoping I'll have some free time to get this wrapped soon!

File: cheroot/server.py [hunk 1]
```diff
@@ -696,7 +696,6 @@ def parse_request(self):
         else:
             if not success:
                 return
-
```
Comment (webknjaz) [inline cheroot/server.py]: This is unrelated.

**Comment by webknjaz** [comment]
@duper51 this might be conflicting with what you're up to

File: cheroot/server.py [hunk 2]
```diff
@@ -1220,15 +1219,13 @@ def communicate(self):
...
-                    # let the conn close.
-                    return
+                    # Allow conn to stay open to support non-blocking sockets
+                    return True
 
                 request_seen = True
                 req.respond()
                 if req.close_connection:
-                    return
+                    return False
```
Comment (jaraco) [inline cheroot/server.py]: The first thing I notice is that these changes seem to be altering the return value of parse_request, to return True or False. I don't see where the meaning of this return value is documented. Can you document it? More importantly, why is it that nothing is using this return value?

File: cheroot/workers/threadpool.py [hunk 3]
```diff
@@ -92,6 +93,53 @@ def __init__(self, server):
...
+            self.bytes_read += self.conn.rfile.bytes_read
+            self.bytes_written += self.conn.wfile.bytes_written
+            self.work_time += time.time() - self.start_time
+            self.start_time = None
+
+    def conn_expired(self, conn, last_active, cur_time):
+        srv_timeout = self.server.timeout
+        return cur_time - last_active > srv_timeout
+
+    def get_expired_conns(self, conn_socks, cur_time):
```
Comment (webknjaz) [inline cheroot/workers/threadpool.py]: I think that creating `cur_time` somewhere separate from actual timeout check is wrong. It should go to `conn_expired` check.

File: cheroot/workers/threadpool.py [hunk 4]
```diff
@@ -92,6 +93,53 @@ def __init__(self, server):
...
+        for conn in conn_list:
+            conn.communicate()  # allow for 408 to be sent
+            conn.close()
+            conn_socks.pop(conn.socket)
+
+    def process_conns(self, conn_socks):
+        rlist = []
+        socks = [sck for sck in conn_socks.keys() if sck.fileno() > -1]
+        if socks:
+            rlist = select.select(socks, [], [], 0)[0]
```
Comment (webknjaz) [inline cheroot/workers/threadpool.py]: 1) why process only first item?
2) what `rlist` stands for? I suggest you use actual words + snake_case for vars

File: cheroot/workers/threadpool.py [hunk 5]
```diff
@@ -92,6 +93,53 @@ def __init__(self, server):
...
+        for sock in rlist:
+            conn, conn_start_time = conn_socks[sock]
+            self.conn = conn
+            self.log_start_stats()
+            try:
+                err_occurred = conn.communicate() is False
+            finally:
+                if err_occurred:
```
Comment (webknjaz) [inline cheroot/workers/threadpool.py]: This will cause `NameError` during the first iteration if communicate errors out and this var never gets initialized.

**Comment by jaraco** [comment]
Thanks for the contrib here. We'll continue to iterate on the effort in #176.

File: cheroot/workers/threadpool.py (hunk 5 above)
Comment (jaraco) [reply cheroot/workers/threadpool.py]: Addressed in 1dd9fd22

File: cheroot/workers/threadpool.py [hunk 6]
```diff
@@ -92,6 +93,53 @@ def __init__(self, server):
...
+
+    def log_close_stats(self):
+        if self.server.stats['Enabled']:
+            self.requests_seen += self.conn.requests_seen
+            self.bytes_read += self.conn.rfile.bytes_read
+            self.bytes_written += self.conn.wfile.bytes_written
+            self.work_time += time.time() - self.start_time
+            self.start_time = None
+
+    def conn_expired(self, conn, last_active, cur_time):
```
[the remark this answers is not in the document]
Comment (jaraco) [reply cheroot/workers/threadpool.py]: Addressed in d83398eb.

File: cheroot/workers/threadpool.py [hunk 7]
```diff
@@ -92,6 +93,53 @@ def __init__(self, server):
         }
         threading.Thread.__init__(self)
 
+    def log_start_stats(self):
+        if self.server.stats['Enabled']:
+            self.start_time = time.time()
+
+    def log_close_stats(self):
+        if self.server.stats['Enabled']:
```
[the remark this answers is not in the document]
Comment (jaraco) [reply cheroot/workers/threadpool.py]: Addressed in 14bcf2a5.

File: cheroot/server.py (hunk 1 above)
Comment (jaraco) [reply cheroot/server.py]: Addressed in 1be10238164f82a7d83bfa8a404a3f1f9961cdf3.
